replication/binlogsyncer.go

```diff
@@ -733,33 +733,53 @@ func (b *BinlogSyncer) parseEvent(s *BinlogStreamer, data []byte) error {
 		// Some events like FormatDescriptionEvent return 0, ignore.
 		b.nextPos.Pos = e.Header.LogPos
 	}
+
+	getCurrentGtidSet := func() GTIDSet {
+		if b.currGset == nil {
+			return nil
+		}
+		return b.currGset.Clone()
+	}
+
+	advanceCurrentGtidSet := func(gtid string) error {
+		if b.currGset == nil {
+			b.currGset = b.prevGset.Clone()
+		}
+		prev := b.currGset.Clone()
+		err := b.currGset.Update(gtid)
+		if err == nil {
+			b.prevGset = prev
+		}
+		return err
+	}
+
 	switch event := e.Event.(type) {
 	case *RotateEvent:
 		b.nextPos.Name = string(event.NextLogName)
 		b.nextPos.Pos = uint32(event.Position)
 		log.Infof("rotate to %s", b.nextPos)
 	case *GTIDEvent:
-		if b.gset == nil {
+		if b.prevGset == nil {
 			break
 		}
 		u, _ := uuid.FromBytes(event.SID)
-		err := b.gset.Update(fmt.Sprintf("%s:%d", u.String(), event.GNO))
+		err := advanceCurrentGtidSet(fmt.Sprintf("%s:%d", u.String(), event.GNO))
 		if err != nil {
 			return errors.Trace(err)
 		}
 	case *MariadbGTIDEvent:
-		if b.gset == nil {
+		if b.prevGset == nil {
 			break
 		}
 		GTID := event.GTID
-		err := b.gset.Update(fmt.Sprintf("%d-%d-%d", GTID.DomainID, GTID.ServerID, GTID.SequenceNumber))
+		err := advanceCurrentGtidSet(fmt.Sprintf("%d-%d-%d", GTID.DomainID, GTID.ServerID, GTID.SequenceNumber))
 		if err != nil {
 			return errors.Trace(err)
 		}
 	case *XIDEvent:
-		event.GSet = b.getGtidSet()
+		event.GSet = getCurrentGtidSet()
 	case *QueryEvent:
-		event.GSet = b.getGtidSet()
+		event.GSet = getCurrentGtidSet()
```

Comment:
`getCurrentGtidSet` is a replacement for  `BinlogSyncer.getGtidSet()`
a. `getCurrentGtidSet` returns `currGset` if set
b. it became local function to emphasize scope of usage